langchain/src/vectorstores/neo4jvector.ts

```diff
@@ -0,0 +1,573 @@
+
+export class Neo4jVectorStore extends VectorStore {
+  private driver: neo4j.Driver;
+
+  private database: string;
+
+  private preDeleteCollection: boolean;
+
+  private nodeLabel: string;
+
+  private embeddingNodeProperty: string;
+
+  private embeddingDimension: number;
+
+  private textNodeProperty: string;
+
+  private keywordIndexName: string;
+
+  private indexName: string;
+
+  private retrievalQuery: string;
+
+  private distanceStrategy: DistanceStrategy = DEFAULT_DISTANCE_STRATEGY;
+
+  _vectorstoreType(): string {
+    return "neo4jvector";
+  }
+
+  constructor(embedings: Embeddings, config: Neo4jVectorStoreArgs) {
+    super(embedings, config);
+  }
+
+  static async initialize(embedings: Embeddings, config: Neo4jVectorStoreArgs) {
+    const store = new Neo4jVectorStore(embedings, config);
+    await store._initializeDriver(config);
+    await store._verifyConnectivity();
+
+    const {
+      preDeleteCollection = false,
+      nodeLabel = "Chunk",
+      textNodeProperty = "text",
+      embeddingNodeProperty = "embedding",
+      keywordIndexName = "keyword",
+      indexName = "vector",
+      retrievalQuery = "",
+    } = config;
+
+    store.embeddingDimension = (await embedings.embedQuery("foo")).length;
+    store.preDeleteCollection = preDeleteCollection;
+    store.nodeLabel = nodeLabel;
+    store.textNodeProperty = textNodeProperty;
+    store.embeddingNodeProperty = embeddingNodeProperty;
+    store.keywordIndexName = keywordIndexName;
+    store.indexName = indexName;
+    store.retrievalQuery = retrievalQuery;
+
+    if (store.preDeleteCollection) {
+      await store._dropIndex();
+    }
+
+    return store;
+  }
+
+  async _initializeDriver({
+    url,
+    username,
+    password,
+    database = "neo4j",
+  }: Neo4jVectorStoreArgs) {
+    try {
+      this.driver = neo4j.driver(url, neo4j.auth.basic(username, password));
+      this.database = database;
+    } catch (error) {
+      throw new Error(
+        "Could not create a Neo4j driver instance. Please check the connection details."
+      );
+    }
+  }
+
+  async _verifyConnectivity() {
+    try {
+      const session = this.driver.session({ database: this.database });
+      await session.close();
+    } catch (error: any) {
+      console.log("Failed to verify connection.");
+    }
+  }
+
+  async _dropIndex() {
+    try {
+      await this.query(`
+        MATCH (n:${this.nodeLabel})
+        CALL {
+          WITH n
+          DETACH DELETE n
+        }
+        IN TRANSACTIONS OF 10000 ROWS;
+      `);
+      await this.query(`DROP INDEX ${this.indexName}`);
+    } catch (error) {
+      console.error("An error occurred while dropping the index:", error);
+    }
+  }
+
+  async query(query: string, params: any = {}): Promise<any[]> {
+    const session = this.driver.session({ database: this.database });
+
+    try {
+      const result = await session.run(query, params);
+      return result.records.map((record: any) => record.toObject());
```

Comment:
This would not handle Neo4j Integers, you'd need to recursively walk through all results, see https://github.com/langchain-ai/langchainjs/blob/efc16f84f94bcea225fb9f306f7db0c59ef03d24/langchain/src/graphs/neo4j_graph.ts#L65

Preferably the actual code for walking through all t records can be moved and reused and t duplicated.